core/internal/filetransfer/file_transfer_s3.go

```diff
@@ -0,0 +1,289 @@
+		digestMismatchError := fmt.Errorf(
+			"digest/etag mismatch: etag %s does not match expected digest %s",
+			*objAttrs.ETag,
+			task.Digest,
+		)
+		if task.VersionId != nil {
+			return nil, ft.formatDownloadError("", digestMismatchError)
+		}
+		getObjInput, ok = ft.getCorrectObjectVersion(bucketName, objectName, task.Digest, getObjInput)
+		if !ok {
+			return nil, ft.formatDownloadError("", digestMismatchError)
+		}
+	}
+	return getObjInput, nil
+}
+
+// listObjectsWithPrefix returns a list of all objects in the specified bucket
+// that begin with the given prefix
+func (ft *S3FileTransfer) listObjectsWithPrefix(
+	bucketName string,
+	prefix string,
+) ([]*s3.GetObjectInput, error) {
+	var objects []*s3.GetObjectInput
+	params := &s3.ListObjectsV2Input{
+		Bucket: aws.String(bucketName),
+		Prefix: aws.String(prefix),
+	}
+	isTruncated := true
+	for isTruncated {
+		output, err := ft.client.ListObjectsV2(ft.ctx, params)
+		if err != nil {
+			return nil, ft.formatDownloadError("error when retrieving objects", err)
+		}
+
+		for _, object := range output.Contents {
+			objects = append(objects, &s3.GetObjectInput{
+				Bucket: aws.String(bucketName),
+				Key:    object.Key,
+			})
+		}
+
+		isTruncated = *output.IsTruncated
+		if isTruncated {
+			params.ContinuationToken = output.NextContinuationToken
+		}
+	}
+	return objects, nil
+}
+
+// getCorrectObjectVersion attempts to find the version of an object with
+// an ETag that matches the specified digest, and returns the input to
+// retrieve that object
+func (ft *S3FileTransfer) getCorrectObjectVersion(
+	bucketName string,
+	objectName string,
+	digest string,
+	getObjInput *s3.GetObjectInput,
+) (*s3.GetObjectInput, bool) {
+	params := &s3.ListObjectVersionsInput{
+		Bucket: aws.String(bucketName),
+		Prefix: aws.String(objectName),
+	}
+	isTruncated := true
+	for isTruncated {
+		versions, err := ft.client.ListObjectVersions(ft.ctx, params)
+		if err != nil {
+			return nil, false
+		}
+		for _, version := range versions.Versions {
+			if trimQuotes(*version.ETag) == digest {
+				getObjInput.Key = aws.String(*version.Key)
+				getObjInput.VersionId = version.VersionId
+				return getObjInput, true
+			}
+		}
+		isTruncated = *versions.IsTruncated
+		if isTruncated {
+			params.KeyMarker = versions.NextKeyMarker
+		}
+	}
+	return nil, false
+}
+
+// downloadFiles downloads all of the objects in the specified bucket
+func (ft *S3FileTransfer) downloadFiles(
+	bucketName string,
+	rootObjectName string,
+	getObjectInputs []*s3.GetObjectInput,
+	basePath string,
+) error {
+	g := new(errgroup.Group)
+	g.SetLimit(maxS3Workers)
+	for _, input := range getObjectInputs {
+		getObjectInput := input // for closure in goroutine
+		g.Go(func() error {
+			objectRelativePath, _ := strings.CutPrefix(*getObjectInput.Key, rootObjectName)
+			localPath := filepath.Join(basePath, filepath.FromSlash(objectRelativePath))
+			err := ft.downloadFile(getObjectInput, localPath)
```

Comment:
You could get away with just returning `ft.downloadFile(...)` and having the format happen at a higher after the errors have been collected. Either way is fine

Reply:
yeah actually I think that I missed some clean up here- I'm going to move all of the formatting to the highest level